[windows] IRGen: add support for DLL Storage semantics

compnerd opened the pull request:
<!-- Please complete this template before creating pull request. -->
#### What's in this pull request?

#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))
<!-- If this pull request resolves any bugs from Swift bug tracker -->

* * * *

<!-- This selection should only be completed by Swift admin -->
Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please smoke test
        OS X platform               | @swift-ci Please smoke test OS X platform
        Linux platform              | @swift-ci Please smoke test Linux platform

 **Validation Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please test
        OS X platform               | @swift-ci Please test OS X platform
        Linux platform              | @swift-ci Please test Linux platform

Note: Only members of the Apple organization can trigger swift-ci.
</details>
<!-- Thank you for your contribution to Swift! -->

Add initial support for modelling DLL Storage semantics for global values.  This
is needed to support the indirect addressing mechanism used on Windows.

compnerd commented:
CC @rjmccall

compnerd commented:
@swift-ci Please test

rjmccall commented on lib/IRGen/GenCast.cpp:
```diff
@@ -329,12 +329,13 @@ static llvm::Function *emitExistentialScalarCastFn(IRGenModule &IGM,
   auto fnTy = llvm::FunctionType::get(returnTy, argTys, /*vararg*/ false);
   auto fn = llvm::Function::Create(fnTy, llvm::GlobalValue::PrivateLinkage,
                                    llvm::Twine(name), IGM.getModule());
+  // FIXME: should we be setting the visibility and the DLL storage as well?
```
No, private things aren't exported.

compnerd replied:
Right, but should actually explicitly set the visibility to hidden and the DLL storage to Default (local only)?

rjmccall replied:
Hidden visibility isn't meaningful for internal / private entities, and I believe LLVM considers it a consistency error to have that combination.  There's no need to make this explicit in the code.

rjmccall commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -561,15 +561,15 @@ void IRGenModule::emitRuntimeRegistration() {
+                             "runtime_registration", getModule());
   RegistrationFunction->setAttributes(constructInitialAttributes());
-  
+  // FIXME: should we be setting the visibility and DLL storage as well?
```
Same thing, private objects aren't exported.

rjmccall commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -926,16 +926,22 @@ void IRGenModule::emitVTableStubs() {
+                                            F.getName(), stub);
+    if (TargetInfo.OutputObjectFormat == llvm::Triple::COFF)
+      alias->setDLLStorageClass(llvm::GlobalValue::DLLExportStorageClass);
+    // FIXME: should we be setting the visibility as well?
```
Default visibility is the default.  I wonder if these really always need to be visible across library boundaries, though.

compnerd replied:
Would it not be a good idea to prepare for something similar to `-fvisibility=hidden`?

rjmccall replied:
LLVM IR does not track whether visibility has been set or not.  Calling the constructor is guaranteed to make the visibility default.

Accordingly, features like -fvisibility=hidden are implemented in frontends by changing the visibility computation.

rjmccall commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -997,15 +1003,14 @@ void IRGenModule::emitTypeVerifier() {
+      fnTy, llvm::GlobalValue::PrivateLinkage, "type_verifier", getModule());
   VerifierFunction->setAttributes(constructInitialAttributes());
-  
+  // FIXME: should we be setting the visibility and DLL storage as well?
```
Private.

rjmccall commented:
The basic approach looks good, but please extrapolate from my comments so far. :)

compnerd commented:
@rjmccall updated based on the extrapolation of your original comments.  Please have another look!

compnerd commented:
@swift-ci Please test

compnerd commented:
@rjmccall post holiday ping :-).

rjmccall commented on lib/IRGen/GenCast.cpp:
```diff
@@ -329,12 +329,12 @@ static llvm::Function *emitExistentialScalarCastFn(IRGenModule &IGM,
                                    llvm::Twine(name), IGM.getModule());
   fn->setAttributes(IGM.constructInitialAttributes());
-  
+
```
I don't actually dislike these whitespace changes, but please try to avoid changing code that you're not actually changing.

rjmccall replied:
This applies to a fair amount of this diff, really.

rjmccall commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -1394,16 +1403,16 @@ llvm::Function *LinkInfo::createFunction(IRGenModule &IGM,
+    IGM.Module.getFunctionList().insert(insertBefore->getIterator(), fn);
+  else
+    IGM.Module.getFunctionList().push_back(fn);
+
```
Why did this code move?

compnerd replied:
That was to have the visibility/calling convention together.  If you would rather that they just be split up around the original position of this, I can do that.

rjmccall commented on lib/IRGen/GenFunc.cpp:
```diff
@@ -1455,11 +1455,11 @@ static llvm::Function *emitBlockDisposeHelper(IRGenModule &IGM,
-                                     IGM.getModule());
+  auto func =
+      llvm::Function::Create(destroyTy, llvm::GlobalValue::InternalLinkage,
+                             "block_destroy_helper", IGM.getModule());
```
Again, please try to avoid reformatting code that you're not changing.

rjmccall commented:
Just some patch-cleanliness requests this time, thanks.

compnerd commented:
Updated.  Sorry, the patch is pretty old, and I was going through and inspecting all the sites that I found, so I had changed code around at one point.  I believe that only the sites where there are changes should now be part of the commit.  Thanks for the patience with this!

rjmccall commented:
LGTM, thanks!

compnerd commented:
@swift-ci Please test and merge
